subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -495,21 +535,26 @@ public boolean executeWithTask(WorkerLease parentWorkerLease, final Action<TaskI
                             @Override
                             public ResourceLockState.Disposition transform(ResourceLockState resourceLockState) {
                                 ResourceLock projectLock = getProjectLock(taskInfo);
+                                if (!projectLock.tryLock()) {
+                                    return FAILED;
+                                }
+                                WorkerLease childLease = parentWorkerLease.createChild();
```

Comment:
I tried acquiring it early, but that actually made it slower, because registering a lock is a pretty expensive operation. 

But even without that, a nice simplification and less garbage.